app/models/queries/projects/selects/life_cycle_step.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+class Queries::Projects::Selects::LifeCycleStep < Queries::Selects::Base
+  KEY = /\Alcsd_(\d+)\z/
+
+  def self.key
+    KEY
+  end
+
+  def self.all_available
+    return [] unless available?
+
+    Project::LifeCycleStepDefinition
+      .pluck(:id)
+      .map { |id| new(:"lcsd_#{id}") }
+  end
+
+  def caption
+    life_cycle.name
+  end
+
+  def life_cycle
+    return @life_cycle if defined?(@life_cycle)
+
+    @life_cycle = Project::LifeCycleStepDefinition
+                    .find_by(id: attribute[KEY, 1])
+  end
```

Comment:
I would suggest calling this method `life_cycle_step_definition`, although it is a bit longer, but there is a distinction between the `life_cycle_step` and the `life_cycle_step_definition`. I'm also just wondering if this declaration has any advantages over using just  a conditional assignment? `||=`

```suggestion
  def life_cycle_step_definition
    @life_cycle_step_definition ||= Project::LifeCycleStepDefinition.find_by(id: attribute[KEY, 1])
  end
```

Reply:
Yes, you are right about the method name. I will adjust it 👍🏻 

About `defined?`, there is actually a slight difference:

1. When we use the `||=` operator here, the right-hand side of the operator (DB query) will be executed as long as the return value is `falsy`. This means when no model can be found and returned, the query would still be executed every time this method is called.
2. When using the `defined?` approach, the DB query is executed exactly once. Even if it returns `nil`, that value is saved in the instance variable and will be returned in future method calls.

I think in this use case it is preferable to not query the DB needlessly, so I went with `defined?`. I usually prefer `||=` because of its elegance, but here it makes sense to invest the extra syntax for `defined?`. What do you think?

Reply:
Thanks for the thorough explanation, yes it makes sense to do that and this is a very good caution agains using `||=` all the time.